File: backend/internal/http/dice.go
package http

import (
	"fmt"
	"math/rand/v2"
	"sort"
	"strings"
)

/*
The dice pool, server side (#176).

A pool is die groups plus one flat modifier — "1d4 + 2d8 + 3" — composed in
the browser and rolled here whenever the roll is made in the open. The server
rolls it rather than accepting a number, because a shared log that takes the
roller's word for the result is not a record of anything.

Mirrors frontend/src/lib/dice.ts, held to it by fixtures/rules/dice-pool.json.
*/

// dieSides are the dice the tower carries. A coin is a d2, because it is.
var dieSides = map[int]bool{2: true, 4: true, 6: true, 8: true, 10: true, 12: true, 20: true, 100: true}

const (
	maxDice     = 100
	maxModifier = 100
)

type dieGroup struct {
	Count int
	Sides int
}

type dicePool struct {
	Groups   []dieGroup
	Modifier int
}

type rolledGroup struct {
	Sides   int
	Results []int
}

type poolResult struct {
	Expression string
	Groups     []rolledGroup
	Modifier   int
	Total      int
	Crit       bool
	Fail       bool
}

// normalizePool merges same-sided groups and drops the empty ones, largest die
// first — so a pool built by tapping d6 four times then twice more reads
// "6d6", and the order of tapping never changes the expression.
func normalizePool(pool dicePool) []dieGroup {
	bySides := map[int]int{}
	for _, g := range pool.Groups {
		if g.Count <= 0 || !dieSides[g.Sides] {
			continue
		}
		bySides[g.Sides] += g.Count
	}
	out := make([]dieGroup, 0, len(bySides))
	for sides, count := range bySides {
		out = append(out, dieGroup{Count: count, Sides: sides})
	}
	sort.Slice(out, func(i, j int) bool { return out[i].Sides > out[j].Sides })
	return out
}

// poolIsRollable reports whether the pool is something the tower will roll.
func poolIsRollable(pool dicePool) bool {
	groups := normalizePool(pool)
	if len(groups) == 0 {
		return false
	}
	dice := 0
	for _, g := range groups {
		dice += g.Count
	}
	if dice > maxDice {
		return false
	}
	if pool.Modifier > maxModifier || pool.Modifier < -maxModifier {
		return false
	}
	return true
}

// diceExpression is how a pool is written: "2d6 + 1d8 + 3". Empty when the
// pool is not rollable, which is also how a caller asks "is this anything?".
// The minus is a true minus sign, matching how the sheet writes them.
func diceExpression(pool dicePool) string {
	if !poolIsRollable(pool) {
		return ""
	}
	parts := make([]string, 0, len(pool.Groups))
	for _, g := range normalizePool(pool) {
		parts = append(parts, fmt.Sprintf("%dd%d", g.Count, g.Sides))
	}
	out := strings.Join(parts, " + ")
	if pool.Modifier > 0 {
		out += fmt.Sprintf(" + %d", pool.Modifier)
	}
	if pool.Modifier < 0 {
		out += fmt.Sprintf(" − %d", -pool.Modifier)
	}
	return out
}

// poolRange is the lowest and highest a pool can land on. ok is false when the
// pool is not rollable.
func poolRange(pool dicePool) (min, max int, ok bool) {
	if !poolIsRollable(pool) {
		return 0, 0, false
	}
	for _, g := range normalizePool(pool) {
		min += g.Count
		max += g.Count * g.Sides
	}
	return min + pool.Modifier, max + pool.Modifier, true
}

// rollPool rolls every die in the pool. ok is false when the pool is not
// rollable — the caller turns that into a 400 rather than an empty roll.
func rollPool(pool dicePool) (poolResult, bool) {
	if !poolIsRollable(pool) {
		return poolResult{}, false
	}
	norm := normalizePool(pool)
	groups := make([]rolledGroup, 0, len(norm))
	total := 0
	for _, g := range norm {
		results := make([]int, 0, g.Count)
		for i := 0; i < g.Count; i++ {
			face := 1 + rand.IntN(g.Sides)
			results = append(results, face)
			total += face
		}
		groups = append(groups, rolledGroup{Sides: g.Sides, Results: results})
	}
	lone20 := len(groups) == 1 && groups[0].Sides == 20 && len(groups[0].Results) == 1
	return poolResult{
		Expression: diceExpression(pool),
		Groups:     groups,
		Modifier:   pool.Modifier,
		Total:      total + pool.Modifier,
		Crit:       lone20 && groups[0].Results[0] == 20,
		Fail:       lone20 && groups[0].Results[0] == 1,
	}, true
}

// diceIn counts the dice actually rolled, across every group.
func diceIn(r poolResult) int {
	n := 0
	for _, g := range r.Groups {
		n += len(g.Results)
	}
	return n
}

// facesOf is every face that came up: "8, 3, 2".
func facesOf(r poolResult) string {
	faces := make([]string, 0, 8)
	for _, g := range r.Groups {
		for _, face := range g.Results {
			faces = append(faces, fmt.Sprint(face))
		}
	}
	return strings.Join(faces, ", ")
}

// rollLine is the sentence the chronicle keeps. The dice are spelled out
// beside the total because "Bramble rolls 27" is a claim, and "8d6: 5, 3, 6,
// 1, 4, 4, 2, 2 = 27" is a roll anyone at the table can check.
//
// Plain text, deliberately: the chronicle prints a message as-is, so markup
// here would reach the feed as literal asterisks.
func rollLine(actor, label string, r poolResult) string {
	var b strings.Builder
	b.WriteString(actor)
	b.WriteString(" rolls ")
	if label != "" {
		b.WriteString(label)
		b.WriteString(" — ")
	}
	b.WriteString(r.Expression)
	b.WriteString(": ")
	b.WriteString(facesOf(r))
	if r.Modifier > 0 {
		b.WriteString(fmt.Sprintf(" + %d", r.Modifier))
	}
	if r.Modifier < 0 {
		b.WriteString(fmt.Sprintf(" − %d", -r.Modifier))
	}
	// One die and nothing added to it is already its own total; "1d20: 18 = 18"
	// reads like the log is showing its working for no reason.
	if diceIn(r) > 1 || r.Modifier != 0 {
		b.WriteString(fmt.Sprintf(" = %d", r.Total))
	}
	switch {
	case r.Crit:
		b.WriteString(" — a natural 20!")
	case r.Fail:
		b.WriteString(" — a natural 1.")
	}
	return b.String()
}

File: frontend/src/lib/dice.ts
/*
The dice pool (#176). A pool is die groups plus one flat modifier — "1d4 +
2d8 + 3" — composed in the moment rather than saved: a combo is what this
spell needs right now, not a thing to name and keep.

Mirrors diceExpression / rollPool in backend/internal/http/dice.go, and
fixtures/rules/dice-pool.json holds the two to the same spelling. That
agreement matters more here than in most mirrors: the browser composes the
pool and the *server* rolls it when the roll is public, so a disagreement
would put one expression on your screen and a different one in the log.
*/

/** The dice the tower carries. A coin is a d2, because it is. */
export const DIE_SIDES = [2, 4, 6, 8, 10, 12, 20, 100] as const;

/** No table rolls more than this at once, and no modifier reaches this far. */
export const MAX_DICE = 100;
export const MAX_MODIFIER = 100;

export interface DieGroup {
  count: number;
  sides: number;
}

export interface DicePool {
  groups: DieGroup[];
  modifier: number;
}

export interface RolledGroup {
  sides: number;
  results: number[];
}

export interface PoolResult {
  expression: string;
  groups: RolledGroup[];
  modifier: number;
  total: number;
  /** A lone d20 landing on 20 or 1 — the only roll the tower calls out. */
  crit: boolean;
  fail: boolean;
}

/**
 * Merge same-sided groups and drop the empty ones, largest die first — so a
 * pool built by tapping d6 four times then twice more reads "6d6", and the
 * order of tapping never changes the expression.
 */
export function normalizePool(pool: DicePool): DieGroup[] {
  const bySides = new Map<number, number>();
  for (const g of pool.groups) {
    if (!Number.isInteger(g.count) || g.count <= 0) continue;
    if (!DIE_SIDES.includes(g.sides as (typeof DIE_SIDES)[number])) continue;
    bySides.set(g.sides, (bySides.get(g.sides) ?? 0) + g.count);
  }
  return [...bySides.entries()]
    .map(([sides, count]) => ({ sides, count }))
    .sort((a, b) => b.sides - a.sides);
}

/** True when the pool is something the tower will actually roll. */
export function poolIsRollable(pool: DicePool): boolean {
  const groups = normalizePool(pool);
  if (groups.length === 0) return false;
  const dice = groups.reduce((n, g) => n + g.count, 0);
  if (dice > MAX_DICE) return false;
  if (!Number.isInteger(pool.modifier) || Math.abs(pool.modifier) > MAX_MODIFIER) return false;
  return true;
}

/**
 * How a pool is written: "2d6 + 1d8 + 3". Empty when the pool is not
 * rollable, which is also how a caller asks "is this anything?".
 *
 * The minus is a true minus sign (−), matching how the sheet already writes
 * negative modifiers.
 */
export function diceExpression(pool: DicePool): string {
  if (!poolIsRollable(pool)) return "";
  const parts = normalizePool(pool).map((g) => `${g.count}d${g.sides}`);
  let out = parts.join(" + ");
  if (pool.modifier > 0) out += ` + ${pool.modifier}`;
  if (pool.modifier < 0) out += ` − ${Math.abs(pool.modifier)}`;
  return out;
}

/** The lowest and highest a pool can land on; null when it is not rollable. */
export function poolRange(pool: DicePool): { min: number; max: number } | null {
  if (!poolIsRollable(pool)) return null;
  const groups = normalizePool(pool);
  return {
    min: groups.reduce((n, g) => n + g.count, 0) + pool.modifier,
    max: groups.reduce((n, g) => n + g.count * g.sides, 0) + pool.modifier,
  };
}

/**
 * Roll a pool here in the browser. This is the private roll — a public one
 * is rolled by the server, so that a feed everyone reads cannot be composed
 * by the client that benefits from it.
 */
export function rollPool(pool: DicePool, rng: () => number = Math.random): PoolResult | null {
  if (!poolIsRollable(pool)) return null;
  const groups = normalizePool(pool).map((g) => ({
    sides: g.sides,
    results: Array.from({ length: g.count }, () => 1 + Math.floor(rng() * g.sides)),
  }));
  const rolled = groups.reduce((n, g) => n + g.results.reduce((a, b) => a + b, 0), 0);
  const lone20 =
    groups.length === 1 && groups[0].sides === 20 && groups[0].results.length === 1;
  return {
    expression: diceExpression(pool),
    groups,
    modifier: pool.modifier,
    total: rolled + pool.modifier,
    crit: lone20 && groups[0].results[0] === 20,
    fail: lone20 && groups[0].results[0] === 1,
  };
}

/** "8, 3, 2" — every face that came up, in the order rolled. */
export function facesOf(result: PoolResult): string {
  return result.groups.flatMap((g) => g.results).join(", ");
}
